libportal/dynamic-launcher.c

```diff
@@ -0,0 +1,609 @@
+                                     "org.freedesktop.portal.DynamicLauncher",
+                                     "Uninstall",
+                                     g_variant_new ("(sa{sv})",
+                                                    desktop_file_id,
+                                                    &opt_builder),
+                                     NULL,
+                                     G_DBUS_CALL_FLAGS_NONE,
+                                     -1,
+                                     NULL, error);
+  return (ret != NULL);
+}
+
+/**
+ * xdp_portal_dynamic_launcher_get_desktop_entry:
+ * @portal: a [class@Portal]
+ * @desktop_file_id: the .desktop file name
+ * @error: return location for a #GError
+ *
+ * This function gets the contents of a .desktop file that was previously
+ * installed by the dynamic launcher portal.
+ *
+ * The @desktop_file_id must be prefixed with the caller's app ID followed by a
+ * "." and suffixed with ".desktop".
+ *
+ * Returns: (transfer full): the contents of the desktop file, or %NULL with
+ *   @error set
+ */
+char *
+xdp_portal_dynamic_launcher_get_desktop_entry (XdpPortal   *portal,
+                                               const char  *desktop_file_id,
+                                               GError     **error)
+{
+  g_autoptr(GVariant) ret = NULL;
+  g_autofree char *contents = NULL;
+
+  g_return_val_if_fail (XDP_IS_PORTAL (portal), NULL);
+  g_return_val_if_fail (desktop_file_id != NULL && *desktop_file_id != '\0', NULL);
+
+  ret = g_dbus_connection_call_sync (portal->bus,
+                                     PORTAL_BUS_NAME,
+                                     PORTAL_OBJECT_PATH,
+                                     "org.freedesktop.portal.DynamicLauncher",
+                                     "GetDesktopEntry",
+                                     g_variant_new ("(s)", desktop_file_id),
+                                     G_VARIANT_TYPE ("(s)"),
+                                     G_DBUS_CALL_FLAGS_NONE,
+                                     -1,
+                                     NULL, error);
+  if (ret == NULL)
+    return NULL;
+
+  g_variant_get (ret, "(s)", &contents);
+  return g_steal_pointer (&contents);
+}
+
+/**
+ * xdp_portal_dynamic_launcher_get_icon:
+ * @portal: a [class@Portal]
+ * @desktop_file_id: the .desktop file name
+ * @out_icon_format: (nullable): return location for icon format string,
+ *   one of "png", "jpeg", "svg"
+ * @out_icon_size: (nullable): return location for icon size
+ * @error: return location for a #GError
+ *
+ * This function gets the icon associated with a .desktop file that was previously
+ * installed by the dynamic launcher portal.
+ *
+ * The @desktop_file_id must be prefixed with the caller's app ID followed by a
+ * "." and suffixed with ".desktop".
+ *
+ * Returns: (transfer full): the icon in a format recognized by g_icon_deserialize(),
+ *   or %NULL with @error set
+ */
+GVariant *
```

Comment:
Shouldn't it return a `GIcon*` instead?

Reply:
I guess it depends how closely we want the libportal API to match the underlying portal API, but I don't see much benefit. Using `GIcon` here and in the `xdp_portal_dynamic_launcher_prepare_install` signature means the caller doesn't have to do `g_icon_serialize`/`g_icon_deserialize`, but it also means libportal has to handle erroring out if the `GIcon` is not a `GBytesIcon` (or let the portal error out instead). Doesn't really seem like enough of a benefit to justify reworking this and the calling code?

Reply:
>it also means libportal has to handle erroring out if the GIcon is not a GBytesIcon (or let the portal error out instead)

I guess we can just replace `g_return_val_if_fail (g_variant_is_of_type (icon_v, G_VARIANT_TYPE ("(sv)")), NULL);` with `g_return_val_if_fail (G_IS_BYTES_ICON (icon), NULL);` when the `GIcon` is being passed in. But then we have to handle setting the error pointer if serialization fails (GLib doesn't set it for us). And similarly we would need to handle setting the error pointer if deserialization fails in `xdp_portal_dynamic_launcher_get_icon()`.

I don't see any use of `g_set_error()` in libportal code aside from in `get_under_snap()`, so I don't know if we want to start using it and make an error domain for libportal (or use something generic like `G_IO_ERROR_FAILED` but that's an anti-pattern). Seems like that would make it a bit harder to reason about when client code is checking error conditions, since the status quo is that errors are only set by portal implementations or at the layer of gdbus, I think.

Another reason it might not make sense to use `GIcon` is in case some clients would want the icon in serialized form when retrieving it with `xdp_portal_dynamic_launcher_get_icon()` or if some clients would already have it serialized and have to deserialize it to pass it as a `GIcon`.

Just some thoughts, not a strong opinion.

Reply:
If this code requires a specific implementation of GIcon, then perhaps it should use that as the accepted type. E.g. `GBytesIcon * xdp_portal_dynamic_launcher_... ()`. Though, on a second thought, using the implementation type directly would limit any future API changes.

Reply:
Perhaps we should stick with `GVariant` then?

Reply:
Let's go with GVariant for now